discovery.md

```diff
@@ -101,18 +101,295 @@ be considered an event subscriber and a target user of this API.
 
 ## API Specification
 
-This document structure focuses on discovery where each document has a two part
-key of `provider` and `type`. This takes a denormalized view where the composite
-key of documents is the `provider` and `type`. Provider is defined as an
-arbitrary human readable string and `type` is the CloudEvent `type` attribute.
+This API is specified as a REST API with well defined entities and relationships
+between those entities.
 
-### Provider entity Attributes
+All responses will have a content-type of `application/json` and are either
+a single JSON encoded object or an array of objects.
+
+### Entities in the API
+
+1. `Provider`
+2. `Type`
+3. `Producer`
+
+The `Provider` and `Type` entities form the basis of the directory and can be
+used to build user experiences around discovery. The `Producer` entity is
+keyed off of `{Provider.Name, Type.Name}` and provides the necessary details to
+create a subscription for events of that `ce-type` from the selected `provider`.
+
+### REST Paths
+
+Each path in the REST API represents either represents a list (or search) over
+and entity class, or the retrieval of an individual entity by name. Each of
+these operations MUST be supported by compliant discovery providers.
+
+#### `/types?matching=[search term]`
+
+Returns a list of all types in the discovery system, optionally filtering on
+a provided search term (`matching`).
+
+##### matching
+
+* Type: `string`
+* Description: Search term that provides case insensitive match against `type`
+  names. The parameter can match any portion of the type name.
+* Constraints:
+  * OPTIONAL
+  * If present, must be non-empty
+* Examples:
+  * `"com.storage.object"`
+    * matches:
+      * `"com.storage.object.create"`
+      * `"com.storage.object.delete"`
+      * `"com.storage.object.update"`
+
+##### Returns
+
+A JSON array of `type` entities.
+
+#### `/types/{name}`
+
+Retrieves an individual type entity by exact match on the `type` name.
```

Comment:
since I can't add a comment down below where I really want to, I'll add it here... I think we'll probably need to be precise on the format of the name for these entities. For example, the list of allowable chars, can it start with a numeric, etc....

Reply:
Agreed, though wouldn't this just follow the format defined in the cloud events spec? https://github.com/cloudevents/spec/blob/v1.0/spec.md#type

Reply:
yup - so even just a pointer to that spec would be ok.